csrc/id_model/predicate_indexing.cpp

```diff
@@ -94,6 +94,37 @@ std::unordered_map<Val*, Val*> getPredicateIndexReplacementMap(
     }
   };
 
+  // If the tensor is circular buffered and the given for-loop is the
+  // main loop of circular buffering, increment the index by
+  // (number_of_stages - 1) since the main loop has a read that is
+  // (number_of_stages - 1) elements ahead.
+  auto replace_for_circular_buffering = [&](ForLoop* fl,
+                                            Val* original_index) -> Val* {
+    auto circular_buffer_axis =
+        GpuLower::current()->circularBufferInfo().getCircularBufferAxis(tv);
+    if (circular_buffer_axis == nullptr ||
+        !id_model.idGraph(IdMappingMode::LOOP)
+             .disjointValSets()
+             .strictAreMapped(fl->iter_domain(), circular_buffer_axis)) {
+      return nullptr;
+    }
+
+    // Epilog should not hit this part
```

Comment:
Is this specifically because there is now a different `IterDomain` mapped to each `ForLoop` that are not Loop mapped with one another and `getCircularBufferAxis(tv)` returns the `IterDomain` of the main loop only?

Reply:
I'll expand the comment here, but that's just because predication of an expr is based on the consumer. If the control reaches here, it means the tensor is circular buffered. In the epilogue loop, the circular buffer tensor should never appear as a consumer but only as a producer. So, what this part means is that this tensor is a consumer tensor and circular buffered, so the loop stage should never be epilogue.

Reply:
Ah ok. That makes sense. Thanks for the explanation.